File: libs/streamlib-deno/clock.ts
// Copyright (c) 2025 Jonathan Fontanez
// SPDX-License-Identifier: BUSL-1.1

/**
 * Canonical monotonic-clock timestamp source for the Deno SDK.
 *
 * Use `monotonicNowNs()` for any timestamp that needs to be compared
 * across processes — frame stamps, escalate request IDs, log
 * correlation tokens, anything that crosses the host/subprocess boundary.
 *
 * Wraps the cdylib's `sldn_monotonic_now_ns()` which calls
 * `clock_gettime(CLOCK_MONOTONIC)` — identical to the syscall Rust's
 * `std::time::Instant::now()` makes on Linux and to what
 * `streamlib-python-native`'s `slpn_monotonic_now_ns()` does. Values
 * from all three sources share the same kernel epoch and are directly
 * comparable.
 *
 * Wall-clock APIs (`Date.now`, `performance.now`) are NOT comparable
 * across processes — `Date.now` drifts under NTP, `performance.now` is
 * relative to each process's `performance.timeOrigin`. Use them only
 * when human-readable wall-clock time is genuinely required (e.g.
 * ISO8601 log formatting).
 */

import type { NativeLib } from "./native.ts";

let _lib: NativeLib | null = null;

/**
 * Wire the native FFI lib for `monotonicNowNs()`. Called by
 * `subprocess_runner` after `loadNativeLib()`. Idempotent.
 */
export function install(lib: NativeLib): void {
  _lib = lib;
}

/** Drop the cached lib reference. Test-only. */
export function _resetForTests(): void {
  _lib = null;
}

/**
 * Current monotonic time in nanoseconds via `clock_gettime(CLOCK_MONOTONIC)`.
 *
 * Comparable across processes on the same kernel — to host Rust
 * `Instant` reads and to the Python SDK's `streamlib.monotonic_now_ns()`.
 *
 * Throws if called before `install()` has wired the native lib —
 * subprocess_runner installs it during startup so processor code can
 * call this freely.
 */
export function monotonicNowNs(): bigint {
  if (!_lib) {
    throw new Error(
      "streamlib clock not installed — call clock.install(lib) before " +
        "monotonicNowNs() (subprocess_runner does this automatically)",
    );
  }
  return _lib.symbols.sldn_monotonic_now_ns() as bigint;
}

File: sdk/streamlib-python-wheel/python/streamlib/clock.py
# Copyright (c) 2025 Jonathan Fontanez
# SPDX-License-Identifier: BUSL-1.1

"""Canonical monotonic-clock timestamp source and drift-free periodic timer.

Use [`monotonic_now_ns`] for any timestamp that needs to be compared across
processes on one machine — frame stamps, log correlation tokens, anything that
crosses a process boundary. It reads the engine's media clock: on Linux
`clock_gettime(CLOCK_MONOTONIC)`, what `time.clock_gettime_ns(time.CLOCK_MONOTONIC)`
reads; on macOS `mach_absolute_time`, what
`time.clock_gettime_ns(time.CLOCK_UPTIME_RAW)` reads, which stops while the
machine sleeps. Every process on one machine, and every stamp the engine
takes, reads the one clock that machine's boot started.

That epoch is the machine's own boot, so a reading from another machine is a
reading of an unrelated clock and subtracting the two means nothing. A bag
that crossed the runtime mesh carries its producer's stamp unchanged; ask
`ctx.inputs.inbound_link_stamp_clock_identity(port, link)` which machine a
link's stamps were taken on before comparing them with another link's, and
[`this_machines_stamp_clock_identity`] which machine the readings taken here
are on before comparing a link's stamps against one of those.

Wall-clock APIs (`time.time`, `datetime.now`, `time.time_ns`) are NOT
comparable across processes — they drift under NTP and reflect different
epochs. Use them only when human-readable wall-clock time is genuinely
required (e.g. ISO8601 log formatting).

[`MonotonicTimer`] is a drift-free periodic timer on that same clock — a
`timerfd` on Linux, a kqueue timer on macOS. The first deadline is
`now + interval` and every one after it is absolute, so ticks never accumulate
drift. Use it as a context manager; `wait(timeout_ms=...)` bounds teardown
latency.
"""

from __future__ import annotations

from ._engine import MonotonicTimer as MonotonicTimer
from ._engine import monotonic_now_ns as monotonic_now_ns
from ._engine import (
    this_machines_stamp_clock_identity as this_machines_stamp_clock_identity,
)

__all__ = [
    "MonotonicTimer",
    "monotonic_now_ns",
    "this_machines_stamp_clock_identity",
]
